core/src/main/java/fr/sncf/osrd/conflicts/IncrementalConflictDetector.kt

```diff
@@ -0,0 +1,150 @@
+package fr.sncf.osrd.conflicts
+
+import com.google.common.collect.Range
+import com.google.common.collect.RangeSet
+import com.google.common.collect.TreeRangeSet
+import fr.sncf.osrd.standalone_sim.result.ResultTrain.SpacingRequirement
+import java.util.TreeMap
+import kotlin.math.max
+import kotlin.math.min
+
+const val DEFAULT_WORK_SCHEDULE_ID: Long = -1
+
+sealed interface IncrementalConflictResponse
+
+data class ConflictResponse(
+    // minimum delay that should be added to the train so that there are no conflicts anymore
+    val minDelayWithoutConflicts: Double,
+    // Time at which the first conflict happened
+    val firstConflictTime: Double,
+) : IncrementalConflictResponse
+
+data class NoConflictResponse(
+    // maximum delay that can be added to the train without creating any conflict
+    val maxDelayWithoutConflicts: Double,
+    // minimum begin time of the next requirement that could conflict
+    val timeOfNextConflict: Double,
+) : IncrementalConflictResponse
+
+fun incrementalConflictDetectorFromTrainReq(
+    requirements: List<TrainRequirements>
+): IncrementalConflictDetector {
+    return IncrementalConflictDetector(convertTrainRequirements(requirements))
+}
+
+fun incrementalConflictDetectorFromReq(
+    requirements: List<Requirements>
+): IncrementalConflictDetector {
+    return IncrementalConflictDetector(requirements)
+}
+
+/**
+ * This class takes a list of requirements as input, and can only be used to compare them to a *new*
+ * set of requirements. The initial requirements cannot be modified. Conflicts between initial
+ * trains are not tested.
+ *
+ * In practice, this is used for STDCM, where the initial requirements represent the timetable
+ * trains, and new requirements come from the train we are trying to fit in the timetable.
+ */
+class IncrementalConflictDetector(requirements: List<Requirements>) {
+    // Zone name -> (end time -> Range(start time, end time))
+    private val spacingZoneUses: Map<String, TreeMap<Double, Range<Double>>>
```

Comment:
This is a bit confusing (redundant?), would it be valid/clearer to move to:
```suggestion
    // Zone name -> (start availability time -> end availability time)
    private val spacingZoneAvailability: Map<String, TreeMap<Double, Double>>
```

Reply:
Or maybe even type the sorted datetime bounds?
```kt
    // Zone name -> sorted (time availability change, becomesAvailable)
    private val spacingZoneAvailability: Map<String, List<Pair<Double, Boolean>>>
```
If we want to dig on the perf here, there may be some room for caching, too: would be useful when time windows are moving on the same path for different requests... The cache could be stable (same infra/STDCM scenario) but quite big.

Reply:
> If we want to dig on the perf here, there may be some room for caching, too: would be useful when time windows are moving on the same path for different requests... The cache could be stable (same infra/STDCM scenario) but quite big (and it requires a shared cache between cores, so valkey would be the minimal option).

In theory yes, but because the requirements use string IDs, checking against the cache takes too long to be faster (I tried). We'd need to fix https://github.com/OpenRailAssociation/osrd/issues/10124 first.

I'll check the other part of the comment soon

Reply:
I don't think the "sorted bounds" would be better, we'd miss out on the `.higherEntry`.

Using availability instead of uses *may* indeed be clearer though. I'll try to make the change and see how it unfolds.

Reply:
> Using availability instead of uses may indeed be clearer though. I'll try to make the change and see how it unfolds.

I tried to make the change but it was too much troubles to make it work. Sometimes we'd end up wanting to do what ends up being `inf - inf`, and figuring out the relevant entry isn't always as straightforward.

Reply:
OK, thanks for trying, maybe adding a little comment on redundancy of `end time` would be nice, I let you resolve the conversation (it's fine without comment too).

Reply:
Commented in this commit: https://github.com/OpenRailAssociation/osrd/pull/9980/commits/77d5703cca4e60f2af93b43afe9c9e553fc3fde5